File: balanced-smileys/balanced.c

#include <stdio.h>
#include <ctype.h>


static int
balanced(char *line)
{
  static char *left[128], *right[128];
  int lstack = 0, rstack = 0;

  for (char *p = line; *p; ++p) {
    switch (*p) {

      case '(':
        left[lstack++] = p;
        break;

      case ')':
        right[rstack++]
    }
  }
}

int
main(int argc, char const *argv[])
{
  FILE *file = fopen(argv[1], "r");
  char line[1024];
  while (fgets(line, 1024, file)) {
    if (strlen(line) > 0) printf("%s\n", balanced(line) ? "YES" : "NO");
  }
  return 0;
}

File: balanced-smileys/balanced.rb
module BalancedSmileys
  def self.analyze_colons(line)
    hash = Hash.new(0)
    line.each_char.each_with_index do |c, i|
      if c == ':'
        hash[:left]  += 1 if line[i-1] == '('
        hash[:right] += 1 if line[i+1] == ')'
      end
    end
    hash
  end

  def self.paren_balance(line)
    balance = 0
    line.each_char do |c|
      case c
      when '('
        balance -= 1
      when ')'
        balance += 1
      end
    end
    balance
  end

  def self.balanced?(line)
    if (balance = paren_balance(line)) == 0
      true
    else
      colons = analyze_colons line
      if balance > 0
        colons[:right] >= balance
      else
        balance *= -1
        colons[:left] >= balance
      end
    end
  end

  def self.run(filename = ARGV[0])
    File.open(filename) do |file|
      file.each_line do |line|
        line.strip!
        next if line.empty?

        puts balanced?(line) ? "YES" : "NO"
      end
    end
  end
end

BalancedSmileys.run if __FILE__ == $0